File: packages/ios-client/Sources/FishjamClient/ReconnectionManager.swift
import Foundation

enum ReconnectionStatus {
    case IDLE
    case RECONNECTING
    case WAITING
    case ERROR
}

public struct ReconnectConfig {
    let maxAttempts: Int
    let initialDelayMs: Int
    let delayMs: Int

    public init(maxAttempts: Int = 5, initialDelayMs: Int = 1000, delayMs: Int = 1000) {
        self.maxAttempts = maxAttempts
        self.initialDelayMs = initialDelayMs
        self.delayMs = delayMs
    }
}

public protocol ReconnectionManagerListener {
    func onReconnectionStarted()
    func onReconnected()
    func onReconnectionRetriesLimitReached()
}

class ReconnectionManager {
    private let reconnectConfig: ReconnectConfig
    private var reconnectAttempts = 0
    var reconnectionStatus = ReconnectionStatus.IDLE
    private let connect: () -> Void
    private var listener: ReconnectionManagerListener

    init(reconnectConfig: ReconnectConfig, connect: @escaping () -> Void, listener: ReconnectionManagerListener) {
        self.reconnectConfig = reconnectConfig
        self.connect = connect
        self.listener = listener
    }

    func onDisconnected() {
        if reconnectAttempts >= reconnectConfig.maxAttempts {
            reconnectionStatus = .ERROR
            listener.onReconnectionRetriesLimitReached()
            return
        }

        if reconnectionStatus == .WAITING {
            return
        }

        if reconnectionStatus != .RECONNECTING {
            reconnectionStatus = .RECONNECTING
            listener.onReconnectionStarted()
        }
        let delay = (reconnectConfig.initialDelayMs + reconnectAttempts * reconnectConfig.delayMs)
        reconnectAttempts += 1

        reconnectionStatus = .WAITING

        DispatchQueue.main.asyncAfter(
            deadline: .now() + DispatchTimeInterval.milliseconds(delay),
            execute: {
                self.reconnectionStatus = .RECONNECTING
                self.connect()
            })
    }

    func onReconnected() {
        if reconnectionStatus != .RECONNECTING {
            return
        }
        reset()
        listener.onReconnected()
    }

    func reset() {
        reconnectAttempts = 0
        reconnectionStatus = .IDLE
    }
}

File: packages/android-client/FishjamClient/src/main/java/com/fishjamcloud/client/ReconnectionManager.kt
package com.fishjamcloud.client

import com.fishjamcloud.client.models.ReconnectConfig
import timber.log.Timber
import java.util.Timer
import kotlin.concurrent.schedule

interface ReconnectionManagerListener {
  fun onReconnectionStarted() {
    Timber.i("Reconnection started")
  }

  fun onReconnected() {
    Timber.i("Reconnected successfully")
  }

  fun onReconnectionRetriesLimitReached() {
    Timber.e("Reconnection retries limit reached")
  }
}

enum class ReconnectionStatus {
  IDLE,
  RECONNECTING,
  ERROR,
  WAITING
}

internal class ReconnectionManager(
  private var reconnectConfig: ReconnectConfig = ReconnectConfig(),
  private val connect: () -> Unit
) {
  private val listeners = mutableListOf<ReconnectionManagerListener>()
  private var reconnectAttempts = 0
  private var reconnectionStatus = ReconnectionStatus.IDLE

  fun onDisconnected() {
    if (reconnectAttempts >= reconnectConfig.maxAttempts) {
      reconnectionStatus = ReconnectionStatus.ERROR
      listeners.forEach { it.onReconnectionRetriesLimitReached() }
      return
    }

    if (reconnectionStatus == ReconnectionStatus.WAITING) {
      return
    }

    if (reconnectionStatus != ReconnectionStatus.RECONNECTING) {
      reconnectionStatus = ReconnectionStatus.RECONNECTING
      listeners.forEach { it.onReconnectionStarted() }
    }
    val delay = reconnectConfig.initialDelayMs + reconnectAttempts * reconnectConfig.delayMs
    reconnectAttempts += 1
    Timer().schedule(delay) {
      reconnectionStatus = ReconnectionStatus.RECONNECTING
      connect()
    }
  }

  fun onReconnected() {
    if (reconnectionStatus != ReconnectionStatus.RECONNECTING) return
    reset()
    listeners.forEach { it.onReconnected() }
  }

  fun reset() {
    reconnectAttempts = 0
    reconnectionStatus = ReconnectionStatus.IDLE
  }

  fun addListener(listener: ReconnectionManagerListener) {
    listeners.add(listener)
  }

  fun removeListener(listener: ReconnectionManagerListener) {
    listeners.remove(listener)
  }
}
